feat: add documentation for go sdk (#501)
matter-labs/zksync-web-era-docs: docs/api/go/accounts-l1-l2.md

# Accounts: L1->L2 transactions

This section explores the methods which allow the [account](./accounts.md) classes to send transactions from L1 to L2.

> If you want to get some background on how L1->L2 interaction works on zkSync, go through the [introduction](../../dev/developer-guides/bridging/l1-l2-interop.md).

Reviewer: ```suggestion
If you want some background on how L1->L2 interaction works on zkSync, go through the [introduction](../../dev/developer-guides/bridging/l1-l2-interop.md).
```

> Full examples of actions below are available at [Getting Started](./getting-started.md) page.

Reviewer: ```suggestion
Full examples of actions below are available on the [getting started](./getting-started.md) page.
```

## Deposit

```go

func (p *DefaultEthProvider) Deposit(token *Token, amount *big.Int, address common.Address, options *GasOptions) (*types.Transaction, error)

```

> **Inputs and Outputs**

Reviewer: ```suggestion
**Inputs and outputs**
```

| Name    | Description                                               |
|---------|-----------------------------------------------------------|
| token   | The address of the deposited L1 ERC20 token.              |
| amount  | The amount of the token to be deposited.                  |
| address | The address that will receive the deposited tokens on L2. |
| options | The options for transaction fee and gas.                  |
| returns | The transaction hash of the deposit.                      |

> ## Claim Failed Deposit

Reviewer: ```suggestion
## Claim failed deposit
```

The `ClaimFailedDeposit` method withdraws funds from the initiated deposit, which failed when finalizing on L2.  

> If the deposit L2 transaction has failed, it sends an L1 transaction calling `ClaimFailedDeposit` method of the 

Reviewer: ```suggestion
If the L2 deposit transaction fails, it sends an L1 transaction calling the `ClaimFailedDeposit` method of the 
```

L1 bridge, which results in returning L1 tokens back to the depositor, otherwise throws the error.

```go

 func (w *Wallet) ClaimFailedDeposit(depositHash common.Hash, ep EthProvider) (common.Hash, error)

```

**Inputs and Outputs**

| Name        | Description                                                 |
|-------------|-------------------------------------------------------------|
| depositHash | The L2 transaction hash of the failed deposit finalization. |
| ep          | The Ethereum provider which interacts with L1 network.      |
| returns     | The transaction hash of claim failed deposit                |

## Withdrawals

Withdrawal are executed in 2 steps: 
 - `Withdraw`: Initiates withdrawal on L2. 
 - `FinalizeWithdraw`: Finalized withdrawal on L1.

### Withdraw

```go
func (w *Wallet) Withdraw(to common.Address, amount *big.Int, token *Token, nonce *big.Int) (common.Hash, error)
```

> **Inputs and Outputs**

Reviewer: ```suggestion
**Inputs and outputs**
```

| Name    | Description                                              |
|---------|----------------------------------------------------------|

> | to      | The address that will receive the withdrew tokens on L1. |

Reviewer: ```suggestion
| to      | The address that receives the withdraw tokens on L1. |
```

| amount  | The amount of the token to be deposited.                 |
| token   | The address of the L2 ERC20 token.                       |
| nonce   | The sender's nonce.                                      |
| returns | The transaction hash of the withdrawal.                  |

> ### FinalizeWithdraw

Reviewer: ```suggestion
### `FinalizeWithdraw`
```

```go
func (w *Wallet) FinalizeWithdraw(withdrawalHash common.Hash, index int) (common.Hash, error)
```

> **Inputs and Outputs**

Reviewer: ```suggestion
**Inputs and outputs**
```

| Name           | Description                                                                          |
|----------------|--------------------------------------------------------------------------------------|
| withdrawalHash | The L2 transaction hash of the withdrawal.                                           |

> | index          | The position in the L2 logs Merkle tree of the l2Log that was sent with the message. |

Reviewer: ```suggestion
| index          | The position in the L2 Merkle-tree logs of the `l2Log` that was sent with the message. |
```

> | returns        | The L1 transaction hash of the finilized withdrawal.                                 |

Reviewer: ```suggestion
| returns        | The L1 transaction hash of the finalized withdrawal.                                 |
```
